File: solutions/cpp/bank-account/1/bank_account.cpp
#include "bank_account.h"
namespace Bankaccount {
  void Bankaccount::open()
  {
    require_closed();
    _balance = 0;
    _open = true;
  }
  void Bankaccount::close()
  {
    require_open();
    _open = false;
  }
  int Bankaccount::balance()
  {
    std::lock_guard guard(_mutex);
    require_open();
    return _balance;
  }
  void Bankaccount::deposit(int amount)
  {
    std::lock_guard guard(_mutex);
    require_open();
    if(amount < 0) {
      throw std::runtime_error("Cannot deposit a negative amount");
    }
    _balance += amount;
  }
  void Bankaccount::withdraw(int amount)
  {
    std::lock_guard guard(_mutex);
    require_open();
    if(amount > _balance) {
      throw std::runtime_error("Account would be overdrawn");
    }
    if(amount < 0) {
      throw std::runtime_error("Cannot withdraw a negative amount");
    }
    _balance -= amount;
  }
  void Bankaccount::require_open()
  {
    if(!_open) {
      throw std::runtime_error("Account must be open to do this");
    }
  }
  void Bankaccount::require_closed()
  {
    if(_open) {
      throw std::runtime_error("Account must be closed to do this");
    }
  }
}

File: solutions/kotlin/bank-account/1/src/main/kotlin/BankAccount.kt
class BankAccount {
    private val lock = Any()
    private var _balance: Long = 0
    private var closed: Boolean = false

    val balance: Long
        get() = synchronized(lock) {
            ensureOpen()
            _balance
        }

    fun adjustBalance(amount: Long) {
        synchronized(lock) {
            ensureOpen()
            _balance += amount
        }
    }

    fun close() {
        synchronized(lock) {
            closed = true
        }
    }

    fun ensureOpen() {
        if (closed) throw IllegalStateException()
    }
}